Demonstrate that #3647 can happen in practice

sfc-gh-anoyes opened the pull request:
Also add assertions and traces that should help catch and diagnose potential txnStateStore bugs.

sfc-gh-anoyes committed:
Assert that mutations that affect txnStateStore are metadata mutations

sfc-gh-anoyes committed:
Assert for now that metadata changes are lock aware

sfc-gh-anoyes committed:
Fix segfault

sfc-gh-anoyes committed:
Add MoveKeysLockFrequently workload

sfc-gh-anoyes committed:
Normalize to one TraceEvent type TxnStateStoreMutation

sfc-gh-anoyes committed:
Explain purpose of class

dongxinEric commented:
This is great!!

jzhou77 commented on fdbserver/ProxyCommitData.actor.h:
```diff
@@ -253,10 +313,11 @@ struct ProxyCommitData {
-	    version(0), minKnownCommittedVersion(0), lastVersionTime(0), commitVersionRequestNumber(1),
-	    mostRecentProcessedRequestNumber(0), getConsistentReadVersion(getConsistentReadVersion), commit(commit),
-	    lastCoalesceTime(0), localCommitBatchesStarted(0), locked(false),
+	    logAdapter(NULL), txnStateStore(&txnStateStoreWrapper), popRemoteTxs(false),
```
Nit:
```suggestion
	    logAdapter(nullptr), txnStateStore(&txnStateStoreWrapper), popRemoteTxs(false),
```

jzhou77 commented on fdbserver/workloads/LockDatabaseFrequently.actor.cpp:
```diff
@@ -0,0 +1,64 @@
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
```
Nit:
```suggestion
 * Copyright 2013-2020 Apple Inc. and the FoundationDB project authors
```

jzhou77 commented on fdbserver/workloads/LockDatabaseFrequently.actor.cpp:
```diff
@@ -0,0 +1,64 @@
+
+	ACTOR static Future<Void> lockAndUnlock(LockDatabaseFrequentlyWorkload* self, Database cx) {
+		state UID uid = deterministicRandom()->randomUniqueID();
+		wait(lockDatabase(cx, uid) && success(delay(self->delayBetweenLocks)));
+		wait(unlockDatabase(cx, uid) && success(delay(self->delayBetweenLocks)));
```
Nit: For the delay to be actually accurate between two lock operations, may be use `self->delayBetweenLocks/2.0` for these two `delay` calls (or even better to randomize the delay of lock and unlock, but keeping the total time constant).

sfc-gh-anoyes replied:
Updated to make delayBetweenLocks closer to what it sounds like it should be and also randomize the delay of lock and unlock

sfc-gh-mpilman commented:
Can this be made against 6.2?

sfc-gh-anoyes committed:
Use nullptr

Co-authored-by: Jingyu Zhou <jingyuzhou@gmail.com>

sfc-gh-anoyes committed:
Update copyright year

Co-authored-by: Jingyu Zhou <jingyuzhou@gmail.com>

sfc-gh-anoyes committed:
Randomize delay between locks using a poisson distribution

Also make the "delayBetweenLocks" parameter more accurate - i.e. if you
have a testDuration of 100 and a delayBetweenLocks of 1, you should
expect to lock about 100 times. If locking or unlocking takes
longer than delayBetweenLocks / 2 then you can't expect as many locks
and it will just kind of lock and unlock as fast as it can.

sfc-gh-anoyes commented:
> Can this be made against 6.2?

I can do that. I still don't have a satisfactory, protocol-compatible fix though. I'm going to keep working on the fix here for now for my convenience. The end goal is to find a protocol-compatible fix and merge it into 6.2.

sfc-gh-mpilman commented:
I think this is a great test, however I don't think this proves that this can happen in practice? I don't think `MoveKeys` issues any non-lockaware transactions (but I could be wrong). The workload itself runs transactions that are not lockaware and change the system keyspace.

sfc-gh-anoyes commented:
> I don't think `MoveKeys` issues any non-lockaware transactions

I think you're right. Usages of move keys within fdbserver open their database in such a way that all transactions created with it are lock aware. Tester's don't make their transactions lock aware, and since I'm calling move keys functions directly with a database from a tester it's not lock aware.

xumengpanda reviewed (commented):
The comments are based on the assumption that this PR will be merged after the problem is fixed.
